mutation_writer/partition_based_splitting_writer.cc

```diff
@@ -20,11 +20,114 @@
  */
 
 #include "mutation_writer/partition_based_splitting_writer.hh"
+#include "memtable.hh"
 
 #include <seastar/core/coroutine.hh>
 
 namespace mutation_writer {
 
+class partition_sorting_mutation_writer {
+    schema_ptr _schema;
+    reader_permit _permit;
+    reader_consumer _consumer;
+    const io_priority_class& _pc;
+    size_t _max_memory;
+    bucket_writer _bucket_writer;
+    std::optional<dht::decorated_key> _last_bucket_key;
+    lw_shared_ptr<memtable> _memtable;
+    future<> _background_memtable_flush = make_ready_future<>();
+    std::optional<mutation> _current_mut;
+    size_t _current_mut_size = 0;
+
+private:
+    void init_current_mut(dht::decorated_key key, tombstone tomb) {
+        _current_mut_size = key.external_memory_usage();
+        _current_mut.emplace(_schema, std::move(key));
+        _current_mut->partition().apply(tomb);
+    }
+
+    future<> flush_memtable() {
+        co_await _consumer(_memtable->make_flush_reader(_schema, _permit, _pc));
```

Comment:
A little disappointing that we need to duplicate this code (and can't just ask the memtable to flush itself), but it's not a big deal. The duplication isn't large and it's still playing with lego bricks.

Reply:
Actually here it is an advantage that flush happens via a reader. The consumer is abstracted, we know it is an sstable, but the code here doesn't.